src/cascadia/UIMarkdown/MarkdownToXaml.cpp

```diff
@@ -0,0 +1,451 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "CodeBlock.h"
+#include "MarkdownToXaml.h"
+
+#include <cmark.h>
+
+namespace winrt
+{
+    namespace MUX = Microsoft::UI::Xaml;
+    namespace WUX = Windows::UI::Xaml;
+    using IInspectable = Windows::Foundation::IInspectable;
+}
+using namespace winrt;
+
+// Bullet points used for unordered lists.
+static const std::wstring bullets[]{
+    L"• ",
+    L"◦ ",
+    L"▪ " // After this level, we'll keep using this one.
+};
+
+template<typename T>
+static std::string_view textFromCmarkString(const T& s) noexcept
+{
+    return std::string_view{ (char*)s.data, (size_t)s.len };
+}
+static std::string_view textFromLiteral(cmark_node* node) noexcept
+{
+    return cmark_node_get_literal(node);
+}
+static std::string_view textFromUrl(cmark_node* node) noexcept
+{
+    return cmark_node_get_url(node);
+}
+
+// Function Description:
+// - Entrypoint to convert a string of markdown into a XAML RichTextBlock.
+// Arguments:
+// - markdownText: the markdown content to render
+// - baseUrl: the current URI of the content. This will allow for relative links
+//   to be appropriately resolved.
+// Return Value:
+// - a RichTextBlock with the rendered markdown in it.
+WUX::Controls::RichTextBlock MarkdownToXaml::Convert(std::string_view markdownText, const winrt::hstring& baseUrl)
+{
+    MarkdownToXaml data{ baseUrl };
+
+    auto doc = cmark_parse_document(markdownText.data(), markdownText.size(), CMARK_OPT_DEFAULT);
+    auto iter = cmark_iter_new(doc);
+    cmark_event_type ev_type;
+    cmark_node* curr;
+
+    while ((ev_type = cmark_iter_next(iter)) != CMARK_EVENT_DONE)
+    {
+        curr = cmark_iter_get_node(iter);
+        data._RenderNode(curr, ev_type);
+    }
+
+    return data._root;
+}
+
+MarkdownToXaml::MarkdownToXaml(const winrt::hstring& baseUrl) :
+    _baseUri{ baseUrl }
+{
+    _root.IsTextSelectionEnabled(true);
+    _root.TextWrapping(WUX::TextWrapping::WrapWholeWords);
+}
+
+WUX::Documents::Paragraph MarkdownToXaml::_CurrentParagraph()
+{
+    if (_lastParagraph == nullptr)
+    {
+        _EndRun(); // sanity check
+        _lastParagraph = WUX::Documents::Paragraph();
+        if (_indent > 0)
+        {
+            if (_indent - _blockQuoteDepth > 0)
+            {
+                _lastParagraph.TextIndent(-12);
+            }
+            _lastParagraph.Margin(WUX::ThicknessHelper::FromLengths(18 * _indent, 0, 0, 0));

```

Comment:
I concur with Dustin that some of these constants are a bit magic. It'd be nice to make them `static constexpr` constants at the top of the file and/or express them "in math" (the way you derived them), because that way we can more easily figure out their meaning in the future once we've already forgotten how this works. 😄 (It would also help with making it configurable if needed.)